blockchain/blockchain.go

```diff
@@ -119,60 +155,56 @@ func Productivity(bc Blockchain, startHeight uint64, endHeight uint64) (map[stri
 	return stats, nil
 }
 
-// blockchain implements the Blockchain interface
-type blockchain struct {
-	mu             sync.RWMutex // mutex to protect utk, tipHeight and tipHash
-	dao            blockdao.BlockDAO
-	config         config.Config
-	blockValidator block.Validator
-	lifecycle      lifecycle.Lifecycle
-	clk            clock.Clock
-	pubSubManager  PubSubManager
-	timerFactory   *prometheustimer.TimerFactory
-
-	// used by account-based model
-	bbf BlockBuilderFactory
-}
-
 // Option sets blockchain construction parameter
-type Option func(*blockchain, config.Config) error
+type Option func(*blockchain) error
 
 // BlockValidatorOption sets block validator
 func BlockValidatorOption(blockValidator block.Validator) Option {
-	return func(bc *blockchain, cfg config.Config) error {
+	return func(bc *blockchain) error {
 		bc.blockValidator = blockValidator
 		return nil
 	}
 }
 
+// PubsubManagerOption creates the pubsub manager
+func PubsubManagerOption(size uint64) Option {
+	return func(bc *blockchain) error {
+		bc.pubSubManager = NewPubSub(size)
+		return nil
+	}
+}
+
 // ClockOption overrides the default clock
 func ClockOption(clk clock.Clock) Option {
-	return func(bc *blockchain, conf config.Config) error {
+	return func(bc *blockchain) error {
 		bc.clk = clk
 		return nil
 	}
 }
 
 // NewBlockchain creates a new blockchain and DB instance
-func NewBlockchain(cfg config.Config, dao blockdao.BlockDAO, bbf BlockBuilderFactory, opts ...Option) Blockchain {
+func NewBlockchain(cfg Config, g genesis.Genesis, dao blockdao.BlockDAO, bbf BlockBuilderFactory, opts ...Option) Blockchain {
 	// create the Blockchain
 	chain := &blockchain{
-		config:        cfg,
-		dao:           dao,
-		bbf:           bbf,
-		clk:           clock.New(),
-		pubSubManager: NewPubSub(cfg.BlockSync.BufferSize),
+		config:  cfg,
+		genesis: g,
+		dao:     dao,
+		bbf:     bbf,
+		clk:     clock.New(),
 	}
 	for _, opt := range opts {
-		if err := opt(chain, cfg); err != nil {
+		if err := opt(chain); err != nil {
 			log.S().Panicf("Failed to execute blockchain creation option %p: %v", opt, err)
 		}
 	}
+	if chain.pubSubManager == nil {
+		chain.pubSubManager = NewPubSub(200)
```

Comment:
have pubsubManager be `nil` if no option is passed?

Reply:
right, if no option, it's `nil`, so create according to default value

Reply:
After offline discussion, the buffersize of pubManager is added into the config of blockchain.